packages/eslint-plugin/src/rules/no-confusing-void-expression.ts

```diff
@@ -388,5 +415,56 @@ export default createRule<Options, MessageId>({
       const type = getConstrainedTypeAtLocation(services, targetNode);
       return tsutils.isTypeFlagSet(type, ts.TypeFlags.VoidLike);
     }
+
+    function isFunctionReturnTypeIncludesVoid(functionType: ts.Type): boolean {
+      const callSignatures = tsutils.getCallSignaturesOfType(functionType);
+
+      return callSignatures.some(signature => {
+        const returnType = signature.getReturnType();
+
+        return tsutils
+          .unionTypeParts(returnType)
+          .some(tsutils.isIntrinsicVoidType);
+      });
+    }
+
+    function isVoidReturningFunctionNode(
+      functionNode:
+        | TSESTree.ArrowFunctionExpression
+        | TSESTree.FunctionDeclaration
+        | TSESTree.FunctionExpression,
+    ): boolean {
+      // Game plan:
+      //   - If the function node has a type annotation, check if it includes `void`.
+      //     - If it does then the function is safe to return `void` expressions in.
+      //   - Otherwise, check if the function is a function-expression or an arrow-function.
+      //   -   If it is, get its contextual type and bail if we cannot.
+      //   - Return based on whether the contextual type includes `void` or not
+
+      const functionTSNode = services.esTreeNodeToTSNodeMap.get(functionNode);
+
+      if (functionTSNode.type) {
+        const returnType = checker.getTypeFromTypeNode(functionTSNode.type);
+
+        return tsutils
+          .unionTypeParts(returnType)
+          .some(tsutils.isIntrinsicVoidType);
+      }
+
+      if (
+        ts.isFunctionExpression(functionTSNode) ||
+        ts.isArrowFunction(functionTSNode)
+      ) {
```

Comment:
[Style] Nit: a little less code to read & run here, as `getContextualType` is just looking for expressions:

```suggestion
      if (ts.isExpression(functionTSNode)) {
```

Reply:
Nice.

Reply:
I've seen [this done here](https://github.com/typescript-eslint/typescript-eslint/blob/ef5df2f83bc20810f5b33ce31f691a8766990977/packages/eslint-plugin/src/rules/no-unsafe-return.ts#L104-L105), would it make sense to send a tiny PR to shorten it in a similar way? Looks like it works out locally.

Reply:
If you want, I'd 👍 it!